File: src/logger.ts
/**
 * Logger implementation
 * Equivalent to Python logger.py with Winston
 */

import winston from 'winston';
import chalk from 'chalk';
import path from 'path';
import fs from 'fs';

// Theme colors matching Python Rich theme
const theme = {
  thought: chalk.cyan,
  command: chalk.bold.white,
  observation: chalk.green,
  error: chalk.bold.red,
  warning: chalk.yellow,
  info: chalk.dim.white,
  step: chalk.bold.magenta,
  header: chalk.bold.blue,
  success: chalk.bold.green,
  blocked: chalk.bold.yellow,
};

export interface LogFunctions {
  logThought: (step: number, thought: string, logger?: winston.Logger) => void;
  logCommand: (step: number, command: string, logger?: winston.Logger) => void;
  logObservation: (step: number, tag: string, status: string, data: string, logger?: winston.Logger) => void;
  logError: (message: string, logger?: winston.Logger) => void;
  logWarning: (message: string, logger?: winston.Logger) => void;
  logHeader: (title: string, subtitle?: string, logger?: winston.Logger) => void;
  logFinal: (answer: string, logger?: winston.Logger) => void;
  logStats: (stats: Record<string, any>, logger?: winston.Logger) => void;
}

export function setupLogger(
  name: string = 'linguclaw',
  logDir: string = 'logs',
  level: string = 'debug'
): winston.Logger {
  const logger = winston.createLogger({
    level,
    defaultMeta: { service: name },
    transports: [],
  });

  // Ensure log directory exists
  if (!fs.existsSync(logDir)) {
    fs.mkdirSync(logDir, { recursive: true });
  }

  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
  
  // File handler
  logger.add(new winston.transports.File({
    filename: path.join(logDir, `session_${timestamp}.log`),
    level: 'debug',
    format: winston.format.combine(
      winston.format.timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
      winston.format.printf(({ level, message, timestamp, ...meta }) => {
        return `${timestamp} | ${level.padEnd(8)} | ${name} | ${message}`;
      })
    ),
  }));

  // Console handler (warnings and above only, to avoid cluttering TUI)
  logger.add(new winston.transports.Console({
    level: 'warning',
    format: winston.format.combine(
      winston.format.colorize(),
      winston.format.simple()
    ),
  }));

  return logger;
}

// Console output functions matching Python Rich behavior
export const logFunctions: LogFunctions = {
  logThought: (step: number, thought: string, logger?: winston.Logger) => {
    console.log(`\n${theme.step(`[Step ${step}]`)} ${theme.thought(`💭 ${thought}`)}`);
    logger?.info(`Step ${step} THOUGHT: ${thought}`);
  },

  logCommand: (step: number, command: string, logger?: winston.Logger) => {
    console.log(`${theme.step(`[Step ${step}]`)} ${theme.command(`⚡ RUN: ${command}`)}`);
    logger?.info(`Step ${step} RUN: ${command}`);
  },

  logObservation: (step: number, tag: string, status: string, data: string, logger?: winston.Logger) => {
    const icons: Record<string, string> = {
      success: theme.success('✓'),
      error: theme.error('✗'),
      blocked: theme.blocked('🚫'),
    };
    const icon = icons[status] || '?';
    console.log(`         ${icon} ${tag}  ${data.slice(0, 120)}`);
    logger?.info(`Step ${step} OBS (${tag}) [${status}]: ${data.slice(0, 200)}`);
  },

  logError: (message: string, logger?: winston.Logger) => {
    console.log(theme.error(`❌ ${message}`));
    logger?.error(message);
  },

  logWarning: (message: string, logger?: winston.Logger) => {
    console.log(theme.warning(`⚠️  ${message}`));
    logger?.warn(message);
  },

  logHeader: (title: string, subtitle: string = '', logger?: winston.Logger) => {
    console.log(theme.header('='.repeat(60)));
    console.log(theme.header(`🤖 ${title}`));
    if (subtitle) {
      console.log(theme.info(subtitle));
    }
    logger?.info(`=== ${title} === ${subtitle}`);
  },

  logFinal: (answer: string, logger?: winston.Logger) => {
    console.log(theme.header('='.repeat(60)));
    console.log(theme.success('FINAL ANSWER'));
    console.log(theme.success('='.repeat(60)));
    console.log(answer);
    console.log(theme.success('='.repeat(60)));
    logger?.info(`FINAL: ${answer}`);
  },

  logStats: (stats: Record<string, any>, logger?: winston.Logger) => {
    console.log(`\n${theme.header('📊 Stats')}`);
    for (const [key, value] of Object.entries(stats)) {
      console.log(`   ${theme.info(`${key}:`)}  ${value}`);
    }
    logger?.info(`Stats: ${JSON.stringify(stats)}`);
  },
};

// Export default logger instance
let defaultLogger: winston.Logger | null = null;

export function getLogger(): winston.Logger {
  if (!defaultLogger) {
    defaultLogger = setupLogger();
  }
  return defaultLogger;
}

export function setDefaultLogger(logger: winston.Logger): void {
  defaultLogger = logger;
}

File: src/logger.py
"""Structured logging with optional Rich colorization."""
import logging
from datetime import datetime
from pathlib import Path
from typing import Optional

try:
    from rich.console import Console
    from rich.theme import Theme
    _RICH = True
except ImportError:
    _RICH = False

if _RICH:
    THEME = Theme({
        "thought": "cyan", "command": "bold white", "observation": "green",
        "error": "bold red", "warning": "yellow", "info": "dim white",
        "step": "bold magenta", "header": "bold blue", "success": "bold green",
        "blocked": "bold yellow",
    })
    console = Console(theme=THEME)
else:
    class _Console:
        def print(self, t: str, **kw): print(t)
        def input(self, p: str) -> str: return input(p)
        def rule(self, t: str = ""): print("=" * 60)
    console = _Console()

def setup_logger(name: str = "linguclaw", log_dir: str = "logs", level: int = logging.DEBUG) -> logging.Logger:
    logger = logging.getLogger(name)
    if logger.handlers: return logger
    logger.setLevel(level)
    Path(log_dir).mkdir(parents=True, exist_ok=True)
    ts = datetime.now().strftime("%Y%m%d_%H%M%S")
    fh = logging.FileHandler(Path(log_dir) / f"session_{ts}.log", encoding="utf-8")
    fh.setLevel(logging.DEBUG)
    fh.setFormatter(logging.Formatter("%(asctime)s | %(levelname)-8s | %(name)s | %(message)s", datefmt="%Y-%m-%d %H:%M:%S"))
    logger.addHandler(fh)
    ch = logging.StreamHandler()
    ch.setLevel(logging.WARNING)
    logger.addHandler(ch)
    return logger

if _RICH:
    def log_thought(s: int, t: str, l: Optional[logging.Logger] = None):
        console.print(f"\n[step]\\[Step {s}][/step] [thought]💭 {t}[/thought]")
        if l: l.info("Step %d THOUGHT: %s", s, t)
    def log_command(s: int, c: str, l: Optional[logging.Logger] = None):
        console.print(f"[step]\\[Step {s}][/step] [command]⚡ RUN: {c}[/command]")
        if l: l.info("Step %d RUN: %s", s, c)
    def log_observation(s: int, t: str, st: str, d: str = "", l: Optional[logging.Logger] = None):
        tag = {"success": "[success]✓", "error": "[error]✗", "blocked": "[blocked]🚫"}.get(st, "?")
        console.print(f"         {tag} {t}[/success]  {d[:120]}")
        if l: l.info("Step %d OBS (%s) [%s]: %s", s, t, st, d[:200])
    def log_error(m: str, l: Optional[logging.Logger] = None):
        console.print(f"[error]❌ {m}[/error]"); 
        if l: l.error(m)
    def log_warning(m: str, l: Optional[logging.Logger] = None):
        console.print(f"[warning]⚠️  {m}[/warning]"); 
        if l: l.warning(m)
    def log_header(t: str, s: str = "", l: Optional[logging.Logger] = None):
        console.rule(f"[header]{t}[/header]"); 
        if l: l.info("=== %s === %s", t, s)
    def log_final(a: str, l: Optional[logging.Logger] = None):
        console.rule("[success]FINAL ANSWER[/success]"); console.print(a); console.rule(); 
        if l: l.info("FINAL: %s", a)
    def log_stats(st: dict, l: Optional[logging.Logger] = None):
        console.print("\n[header]📊 Stats[/header]"); 
        for k, v in st.items(): console.print(f"   [info]{k}:[/info]  {v}")
        if l: l.info("Stats: %s", st)
else:
    def log_thought(s: int, t: str, l: Optional[logging.Logger] = None):
        print(f"\n[Step {s}] 💭 {t}"); 
        if l: l.info("Step %d THOUGHT: %s", s, t)
    def log_command(s: int, c: str, l: Optional[logging.Logger] = None):
        print(f"[Step {s}] ⚡ RUN: {c}"); 
        if l: l.info("Step %d RUN: %s", s, c)
    def log_observation(s: int, t: str, st: str, d: str = "", l: Optional[logging.Logger] = None):
        icon = {"success": "✓", "error": "✗", "blocked": "🚫"}.get(st, "?")
        print(f"         {icon} {t}: {d[:120]}")
        if l: l.info("Step %d OBS (%s) [%s]: %s", s, t, st, d[:200])
    def log_error(m: str, l: Optional[logging.Logger] = None):
        print(f"❌ {m}"); 
        if l: l.error(m)
    def log_warning(m: str, l: Optional[logging.Logger] = None):
        print(f"⚠️  {m}"); 
        if l: l.warning(m)
    def log_header(t: str, s: str = "", l: Optional[logging.Logger] = None):
        print("=" * 60); print(f"🤖 {t}"); 
        if l: l.info("=== %s === %s", t, s)
    def log_final(a: str, l: Optional[logging.Logger] = None):
        print("\n=== FINAL ANSWER ==="); print(a); print("=" * 60); 
        if l: l.info("FINAL: %s", a)
    def log_stats(st: dict, l: Optional[logging.Logger] = None):
        print("\n=== Stats ==="); 
        for k, v in st.items(): print(f"   {k}: {v}")
        if l: l.info("Stats: %s", st)
